pkg/controller/container-runtime-config/helpers.go

```diff
@@ -0,0 +1,229 @@
+		)
+	} else {
+		condition = mcfgv1.NewContainerRuntimeConfigCondition(
+			mcfgv1.ContainerRuntimeConfigSuccess,
+			v1.ConditionTrue,
+			"Success",
+		)
+	}
+	if len(args) > 0 {
+		format, ok := args[0].(string)
+		if ok {
+			condition.Message = fmt.Sprintf(format, args[:1]...)
+		}
+	}
+	return *condition
+}
+
+// updateStorageConfig decodes the data rendered from the template, merges the changes in and encodes it
+// back into a TOML format. It returns the bytes of the encoded data
+func updateStorageConfig(data []byte, internal *mcfgv1.ContainerRuntimeConfiguration) ([]byte, error) {
+	tomlConf := new(tomlConfigStorage)
+	if _, err := toml.Decode(string(data), tomlConf); err != nil {
+		return nil, fmt.Errorf("error unmarshalling crio config: %v", err)
+	}
+
+	if internal.OverlaySize != (resource.Quantity{}) {
+		tomlConf.Storage.Options.Size = internal.OverlaySize.String()
+	}
+
+	var newData bytes.Buffer
+	encoder := toml.NewEncoder(&newData)
+	if err := encoder.Encode(*tomlConf); err != nil {
+		return nil, err
+	}
+
+	return newData.Bytes(), nil
+}
+
+// updateCRIOConfig decodes the data rendered from the template, merges the changes in and encodes it
+// back into a TOML format. It returns the bytes of the encoded data
+func updateCRIOConfig(data []byte, internal *mcfgv1.ContainerRuntimeConfiguration) ([]byte, error) {
+	tomlConf := new(tomlConfigCRIO)
+	if _, err := toml.Decode(string(data), tomlConf); err != nil {
+		return nil, fmt.Errorf("error unmarshalling crio config: %v", err)
+	}
+
+	if internal.PidsLimit > 0 {
+		tomlConf.Crio.Runtime.PidsLimit = internal.PidsLimit
+	}
+	if internal.LogSizeMax != (resource.Quantity{}) {
+		tomlConf.Crio.Runtime.LogSizeMax = internal.LogSizeMax.Value()
+	}
+	if internal.LogLevel != "" {
+		tomlConf.Crio.Runtime.LogLevel = internal.LogLevel
+	}
+	if internal.InfraImage != "" {
+		tomlConf.Crio.Image.PauseImage = internal.InfraImage
+	}
+
+	var newData bytes.Buffer
+	encoder := toml.NewEncoder(&newData)
+	if err := encoder.Encode(*tomlConf); err != nil {
+		return nil, err
+	}
+
+	return newData.Bytes(), nil
+}
+
+// validateUserContainerRuntimeConfig ensures that the values set by the user are valid
+func validateUserContainerRuntimeConfig(cfg *mcfgv1.ContainerRuntimeConfig) error {
+	if cfg.Spec.ContainerRuntimeConfig == nil {
+		return nil
+	}
+	ctrcfgValues := reflect.ValueOf(*cfg.Spec.ContainerRuntimeConfig)
+	if !ctrcfgValues.IsValid() {
+		return fmt.Errorf("containerRuntimeConfig is not valid")
+	}
+
+	ctrcfg := cfg.Spec.ContainerRuntimeConfig
+	if ctrcfg.PidsLimit > 0 && ctrcfg.PidsLimit < minPidsLimit {
+		return fmt.Errorf("invalid PidsLimit %q, cannot be less than 20", ctrcfg.PidsLimit)
+	}
+
+	if ctrcfg.LogSizeMax.Value() > 0 && ctrcfg.LogSizeMax.Value() <= minLogSize {
+		return fmt.Errorf("invalid LogSizeMax %q, cannot be less than 8kB", ctrcfg.LogSizeMax.String())
+	}
+
+	validLogLevels := []string{"error", "fatal", "panic", "warn", "info", "debug"}
```

Comment:
can you use  a `map[string]bool` to avoid the loop below?

Reply:
and move `validLogLevels` inside the `if`

Reply:
Changed to a map. I chose string initially so I can easily print the valid values when an error is returned. But map works too :)